lib/Runtime/HostManager/HostManager.cpp

```diff
@@ -344,37 +383,42 @@ HostManager::runNetwork(llvm::StringRef networkName,
         std::move(context));
     return currentRun;
   }
-
+  // Setup the request
+  InferRequest queuedRequest;
+  queuedRequest.networkName = networkName;
+  queuedRequest.context = std::move(context);
+  queuedRequest.priority = priority;
+  queuedRequest.callback = callback;
+  queuedRequest.requestID = currentRun;
+  // If we haven't reached maxActuveRequests kick off request.
   size_t activeRequestCount = activeRequestCount_++;
-  if (activeRequestCount >= config_.maxActiveRequests) {
-    activeRequestCount_--;
+  if (activeRequestCount < config_.maxActiveRequests) {
+    dispatchRun(std::move(queuedRequest));
+    return currentRun;
+  }
+  // We are at max inflights, next we check if there is space in the priority
+  // queue and return an error if we are out of space.
+  size_t queueSize = queuedRequestCount_++;
+  activeRequestCount_--;
+  if (queueSize >= config_.maxQueueSize) {
+    // The queue is full, return an error.
+    queuedRequestCount_--;
     network->refcount--;
     callback(
         currentRun,
         MAKE_ERR(GlowErr::ErrorCode::RUNTIME_REQUEST_REFUSED,
-                 strFormat("The number of allowed requests has been exceeded. "
-                           "active requests: %lu allowed requests: %zu",
-                           activeRequestCount, config_.maxActiveRequests)),
+                 strFormat(
+                     "The number of allowed queued requests has been exceeded. "
+                     "queued requests: %lu allowed requests: %zu",
+                     queueSize, config_.maxQueueSize)),
         std::move(context));
     return currentRun;
   }
+  // We have space in the queue, queue up the request for later.
 
-  executor_->run(networks_[networkName].dag.root.get(), std::move(context),
-                 currentRun,
-                 [this, callback, name = networkName.str()](
-                     RunIdentifierTy runID, llvm::Error err,
-                     std::unique_ptr<ExecutionContext> context) {
-                   {
-                     std::lock_guard<std::mutex> networkLock(networkLock_);
-                     auto it = networks_.find(name);
-                     if (it != networks_.end()) {
-                       it->second.refcount--;
-                     }
-                   }
-                   TRACE_EVENT_INSTANT(context->getTraceContext(),
-                                       TraceLevel::RUNTIME, "finish_" + name);
-                   callback(runID, std::move(err), std::move(context));
-                   --activeRequestCount_;
-                 });
+  {
+    std::lock_guard<std::mutex> networkLock(networkLock_);
+    inferQueue_.push(std::move(queuedRequest));
```

Comment:
So I think it might be possible to drop requests where we're close to the active request limit.  Here's what I suspect can happen:
Thread 1: request1 comes in, less than maxActive, dispatchRun is called, submits r1 to executor (T3)
Thread 2: request2 comes in, greater than maxActive, skips dispatchRun, but before it can be enqueued...
Thread 3 (executor): processes request1, sees an empty queue, returns
Thread 2: enqueues r2
... and nothing ever calls dispatchRun to process r2.

In fact it seems like the service could deadlock, if a bunch of requests pile up between the executor and the enqueue.

The control flow and synchronization here is pretty confusing so maybe this can't happen... but can you convince me it's not possible?

Reply:
I think you're right, the second group of requests would be stuck until another request comes in, I'll fix it.

Reply:
I'm just skimming this, but that makes sense to me. Basically if thread2 gets descheduled at HostManager.cpp::399 while thread3 finishes the first request, nothing will pick up its request.
Also I agree that this is a bit confusing. The atomic state shared between the mutexed queues and the std::atomic counters describing the queues is a bit hard to follow.

Reply:
I think some of this complexity is I was trying to avoid enqueuing a request if it could be immediately dispatched. I think if I add the request to the queue first then there should be no race condition. 
T1 R1 comes in is enqueued, is dispatched
T2 R2, gets enqueued, if T1 finishes and grabs R2, then maxActive will still be at the limit, if T1 finishes early then maxActive will drop and R2 will get dispatched.
This means some amount of increased latency but I think that's better than more locking.